src/backend/optimizer/plan/createplan.c

```diff
@@ -667,12 +667,27 @@ disuse_physical_tlist(PlannerInfo *root, Plan *plan, Path *path)
 			plan->targetlist = build_path_tlist(root, path);
 
 			/**
-			 * If plan has a flow node, ensure all entries of hashExpr
-			 * are in the targetlist.
+			 * If plan has a flow node, ensure all entries of hashExpr,
+			 * which are not of type Param, are in the targetlist.
 			 */
 			if (plan->flow && plan->flow->hashExprs)
 			{
-				plan->targetlist = add_to_flat_tlist_junk(plan->targetlist, plan->flow->hashExprs, true /* resjunk */);
+				List		*val_exprs = NIL;
+				ListCell	*lc;
+
+				foreach(lc, plan->flow->hashExprs)
+				{
+					Expr	*expr = (Expr *) lfirst(lc);
+
+					if (!IsA(expr, Param))
```

Comment:
> The hashExprs list consists of either distribution key or it's EquivalenceClass members, which could appear with WHERE clause.

hashExpr of Flow [must contain](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/cdb/cdbpathlocus.c#L365-L377) only distribution keys references for scan nodes, represented as references to [equivalence class](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/README#L397) lists. As far I understand, equivalence classes are built not only for where clauses, but for join condition too. Do we have guaranties of filtering other relation fields?

Reply:
I've recheck existing solution for query from tests with debugger:

```
Thread 1 "postgres" hit Breakpoint 3, disuse_physical_tlist (root=0x55e5b9b38308, plan=0x55e5b9adfa88, path=0x55e5b9a00b30) at createplan.c:667
(gdb) n
(gdb) p nodeToString(plan->flow->hashExprs)
$2 = 0x55e5b9ae0d18 "({PARAM :paramkind 1 :paramid 0 :paramtype 25 :paramtypmod -1 :paramcollid 100 :location 83})"
```

`hashExprs` contains only PARAM entry, not whole Equivalence class. Am I correct? In this case, I think that it's an invalid hash entry value, and we must research the reason of building such value. As far I understand, new solutions tries to fix this problem.